### DISCUSSION TITLE
Preserving `parent` during Transform

### DISCUSSION THREAD

**Discussion by amoffat** [lark-parser/lark#1333] [Q&A]
Hi there, I'm trying to attach a `.parent` attribute to each `Tree` in the parsed result, as explained with this example:

https://lark-parser.readthedocs.io/en/latest/recipes.html#keeping-track-of-parents-when-visiting

However, when my tree is processed by a `Transformer`, the `.parent` attribute no longer exists. Here is a minimal example, setting an `.id` attribute and then checking for that attribute in the `Transformer`:

```python
from uuid import uuid4

from lark import Lark, Transformer, Tree, Visitor

grammar = Lark(
    maybe_placeholders=False,
    grammar=r"""
    start: expr -> calculation
    expr: term ("+" term)*
    term: factor ("*" factor)*
    factor: NUMBER | "(" expr ")"
    %import common.NUMBER
    %ignore " "
    """,
)


class AddParent(Visitor):
    def __default__(self, node):
        node.id = uuid4()


class MyTransformer(Transformer):
    def term(self, items):
        assert hasattr(items[0], "id")
        return Tree("factor", items)


tree = grammar.parse("1 + 2 * 3")
AddParent().visit(tree)
MyTransformer().transform(tree)
```

If I set the attribute on the `Tree.meta` property, it works, but it doesn't work on the `Tree` object itself. I have also tried using `Transformer_InPlace`, with the hopes that the original `Transformer` was making a stripped copy of each `Tree` object, but the result is the same: each transform callback method cannot see the `.id` property.

Any idea what I am doing wrong?

**Comment by erezsh** [comment]
Sorry, I'm only seeing this now.

You're right that the transform creates new tree nodes, which would be without the 'id' attribute.

However, using Transformer_InPlace should work.

If you're still having trouble with it, let me know, and I'll try to guide you towards a solution.
